### ISSUE TITLE
bug: Trouble spams error messages under certain circumstances

### ISSUE THREAD

**Issue by KodyVB** [folke/trouble.nvim#527]
### Did you check docs and existing issues?

- [X] I have read all the trouble.nvim docs
- [X] I have searched the existing issues of trouble.nvim
- [X] I have searched the existing issues of plugins related to this issue

### Neovim version (nvim -v)

NVIM v0.10.0

### Operating system/version

Arch Linux 6.9.7

### Describe the bug

Using Ruff and Python, if the wrong indentation is used and Trouble is brought up, a concat error is thrown, usually many times.

### Steps To Reproduce

1. Create a Python file with an error which will not cause this issue and save it
```py
def main() -> None:
    prnt('a')
    print('b')


if __name__ == "__main__":
    main()
```
2. Run `:lua require("trouble").open("diagnostics")`
3. With the diagnostic tab still open, change the Python file so it has no errors and save it again
```py
def main() -> None:
    print('a')
    print('b')


if __name__ == "__main__":
    main()
```
4. Now introduce an indentation error
```py
def main() -> None:
        print('a')
    print('b')


if __name__ == "__main__":
    main()
```
5. Move the cursor around in normal mode and open command mode via `:`

### Expected Behavior

Any errors should ideally show up once - this causes continuous errors after the cursor is moved around a few times.

### Repro

```Lua
-- DO NOT change the paths and don't remove the colorscheme
local root = vim.fn.fnamemodify("./.repro", ":p")

-- set stdpaths to use .repro
for _, name in ipairs({ "config", "data", "state", "cache" }) do
  vim.env[("XDG_%s_HOME"):format(name:upper())] = root .. "/" .. name
end

-- bootstrap lazy
local lazypath = root .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", lazypath, })
end
vim.opt.runtimepath:prepend(lazypath)

-- install plugins
local plugins = {
  "folke/tokyonight.nvim",
  "folke/trouble.nvim",
{
-- for getting ruff installed
  "williamboman/mason.nvim",
  build = ":MasonUpdate",
  dependencies = {
    "WhoIsSethDaniel/mason-tool-installer.nvim",
  },
  config = function()
    local mason = require("mason")
    local mason_tool_installer = require("mason-tool-installer")

    mason.setup({})

    mason_tool_installer.setup({
      ensure_installed = {
        "ruff",
      },
      auto_update = false,
      run_on_start = true,
      start_delay = 0,
      debounce_hours = 0,
    })
  end,
},
{
-- to get ruff linting
  "mfussenegger/nvim-lint",
  config = function()
    local lint = require("lint")
    lint.linters_by_ft = {
      python = { "ruff" },
    }

    local lint_augroup = vim.api.nvim_create_augroup("lint", { clear = true })

    vim.api.nvim_create_autocmd({ "BufEnter", "BufWritePost", "TextChanged" }, {
      group = lint_augroup,
      callback = function()
        lint.try_lint()
      end,
    })
  end,
}

}
require("lazy").setup(plugins, {
  root = root .. "/plugins",
})

vim.cmd.colorscheme("tokyonight")
-- add anything else here
```

**Comment by folke** [comment]
Something weird is going on with ruff and Neovim.
It generates a `vim.NIL` as the diagnostics code, whcih is not supposed to happen.

Fixed it anyway. Thank you for the detailed bug report!

**Comment by seankim658** [comment]
FWIW I'm on the latest commit and started getting this error. I found this ticket and although (at least I'm pretty sure) I don't have any indentation issues I get this. I disabled ruff (also first tried updating ruff to no avail) and trouble works fine again, so think ruff might still be having issues?

<img width="1331" alt="image" src="https://github.com/user-attachments/assets/0161b088-c4cd-4615-8065-84b4b4b154a8" />
